### PR TITLE
Bugfix: Wallet: Soft-fail exceptions within ListWalletDir file checks

### PR DESCRIPTION
**Pull request by luke-jr**
Previously, an exception would be thrown, which could kill the node in some circumstances.

Includes test changes to cause failure.

Review with `?w=1`

### CODE REVIEW

File: test/functional/wallet_multiwallet.py [hunk 1]
```diff
@@ -50,6 +50,11 @@ def wallet_file(name):
         os.mkdir(wallet_dir('w7'))
         os.symlink('w7', wallet_dir('w7_symlink'))
 
+        os.symlink('..', wallet_dir('recursive_dir_symlink'))
+
+        os.mkdir(wallet_dir('self_walletdat_symlink'))
+        os.symlink('wallet.dat', wallet_dir('self_walletdat_symlink/wallet.dat'))
```
Comment (jonatack) [inline test/functional/wallet_multiwallet.py:56]: Verified that this does cause the test to fail at line 79 without the changes in this PR.

Could also assert that the new logging is taking place with this at line 79 or just after:
```python3
with self.nodes[0].assert_debug_log(expected_msgs=['Too many levels of symbolic links', 'Error scanning']):
    self.nodes[0].listwalletdir()['wallets']
```

**Review by jonatack — COMMENTED** [review]
ACK bca838d38c54f3df3c85da54e23857f06f07c6bb

File: test/functional/wallet_multiwallet.py (hunk 1 above)
Comment (jonatack) [reply test/functional/wallet_multiwallet.py:56]: (or combined with the existing line 79 to share the same listwalletdir call)
```diff
-        assert_equal(sorted(map(lambda w: w['name'], self.nodes[0].listwalletdir()['wallets'])), ['', os.path.join('sub', 'w5'), 'w', 'w1', 'w2', 'w3', 'w7', 'w7_symlink', 'w8'])
+        with self.nodes[0].assert_debug_log(expected_msgs=['Too many levels of symbolic links', 'Error scanning']):
+            list = self.nodes[0].listwalletdir()['wallets']
+        assert_equal(sorted(map(lambda w: w['name'], list)), ['', os.path.join('sub', 'w5'), 'w', 'w1', 'w2', 'w3', 'w7', 'w7_symlink', 'w8'])
```

**Comment by Aonny112** [comment]
Help me.

**Comment by jonatack** [comment]
ACK 4f0cbc4

**Review by meshcollider — COMMENTED** [review]
Code review ACK 4f0cbc4bc74dd87bc187fb839218b1fa64cfb205

**Review by hebasto — APPROVED** [review]
I've reviewed 4f0cbc4bc74dd87bc187fb839218b1fa64cfb205 and ready to ACK after rebasing.

**Comment by luke-jr** [comment]
Rebased

**Review by hebasto — APPROVED** [review]
ACK 1e77a8d5a4faa0757e53e0284662b249eebc3ef7, I have reviewed the code and it looks OK, I agree it can be merged.

**Review by jonatack — COMMENTED** [review]
re-ACK 1e77a8d5a4faa0757e53e0284662b249eebc3ef7

In the new test, the new exception prints:

[httpworker.1] ListWalletDir: Error scanning /tmp/bitcoin_func_test_ql5k4c0f/node0/regtest/wallets/self_walletdat_symlink: boost::filesystem::status: Too many levels of symbolic links: "/tmp/bitcoin_func_test_ql5k4c0f/node0/regtest/wallets/self_walletdat_symlink/wallet.dat"

[httpworker.1] ListWalletDir: Error scanning /tmp/bitcoin_func_test_ql5k4c0f/node0/regtest/wallets/self_walletdat_symlink/wallet.dat: boost::filesystem::status: Too many levels of symbolic links: "/tmp/bitcoin_func_test_ql5k4c0f/node0/regtest/wallets/self_walletdat_symlink/wallet.dat"`

**Comment by Saibato** [comment]
nit: Before merge please note; this  ( #19502 and  ( #20080  or #19933 ) should merged together, since they vice versa would create new issues, like shorted wallet names.

If we have boost cononical best first #20080 and then this.

**Comment by luke-jr** [comment]
I don't see what new issues (aside from rebasing) would be created by merging them separately?

**Comment by Saibato** [comment]
> I don't see what new issues (aside from rebasing) would be created by merging them separately?

iirc this a fix for the boost link problem and  ( evil wallet ) and makes #18095 obsolete.  

But if this get merged first, then real short filenames wallets from the later found  bug  could be created from a bogus dirname on disk for real.

**Review by meshcollider — COMMENTED** [review]
re-utACK 1e77a8d5a4faa0757e53e0284662b249eebc3ef7

**Comment by Saibato** [comment]
> But if this get merged first, then real short filenames wallets from the later found bug could be created from a bogus dirname on disk for real.

@luke-jr 
Rechecked complies from the sources:  To open non exitsting files,  is no longer the case with current master and knots so no merge sequence preferences here any longer. Was only a problem short after those errors where found and exits only if direct on top of last releases before/= 0.20

**Comment by luke-jr** [comment]
Rebased again

**Review by hebasto — APPROVED** [review]
~re-ACK 9f74b7b3beba182d3b3118bf331515ef6e5ffcf2~, only rebased since my [previous](https://github.com/bitcoin/bitcoin/pull/19502#pullrequestreview-506748519) review, verified with `git range-diff master 1e77a8d5a 9f74b7b3b`.

Revoked. See https://github.com/bitcoin/bitcoin/pull/19502#issuecomment-713165606

**Comment by hebasto** [comment]
I revoke my ACK as this change makes possible an infinity loop.
To reproduce the bug, one could create a directory without access permissions, e.g.:
```
$ ls -l /home/hebasto/bitcoin-data-dir/wallets
total 1456
drwx------ 3 hebasto hebasto    4096 Oct  3 11:16 20191109-main
drwx------ 2 root    root       4096 Oct 20 23:58 crash
-rw-r--r-- 1 hebasto hebasto       0 Mar  5  2020 db.log
drwx------ 3 hebasto hebasto    4096 Oct 20 23:56 w20201015.bdb
drwx------ 3 hebasto hebasto    4096 Oct 20 23:56 w20201015.desc
-rw-r----- 1 hebasto hebasto 1474560 Sep 12 16:08 wallet.dat
```

I suggest to combine into this PR the approach from #18095: https://github.com/bitcoin/bitcoin/blob/6904a309194c7ddf5a071d43e6c2bf892e8fe396/src/wallet/walletutil.cpp#L42-L53
with credits to @uhliksk

**Comment by luke-jr** [comment]
@hebasto Couldn't reproduce, but it did cause listwallets to miss other wallets. Added a fix and test.

**Comment by hebasto** [comment]
Tested 49cc6919c4f9ef5a33ad5948a1b43496b454fab0, still able to reproduce an infinite loop:
```
cd <path-to-wallet-dir>
sudo mkdir crash
sudo chown 0700 crash
```

then start a node, and initiate a `ListWalletDir` call in any way.

**Review by hebasto — COMMENTED** [review]
Approach ACK 44c0bc31f4fadc932877d9764402644c84f3be4c, tested on Linux Mint 20 (x86_64).

Instead of an [infinite loop](https://github.com/bitcoin/bitcoin/pull/19502#issuecomment-719004065) having a nice log message:
```
2020-10-29T20:41:26Z [main] ListWalletDir: Error scanning /home/hebasto/.bitcoin/testnet3/wallets/crash: boost::filesystem::status: Permission denied: "/home/hebasto/.bitcoin/testnet3/wallets/crash/wallet.dat"
```

Keeping reviewing the last commit.

File: test/functional/wallet_multiwallet.py [hunk 2]
```diff
@@ -109,7 +115,16 @@ def wallet_file(name):
             self.nodes[0].createwallet(wallet_name, descriptors=False)
         for wallet_name in wallet_names[-2:]:
             self.nodes[0].loadwallet(wallet_name)
-        assert_equal(sorted(map(lambda w: w['name'], self.nodes[0].listwalletdir()['wallets'])), [self.default_wallet_name, os.path.join('sub', 'w5'), 'w', 'w1', 'w2', 'w3', 'w7', 'w7_symlink', 'w8'])
+
+        os.mkdir(wallet_dir('no_access'))
+        os.chmod(wallet_dir('no_access'), 0)
```
Comment (hebasto) [inline test/functional/wallet_multiwallet.py:120]: 44c0bc31f4fadc932877d9764402644c84f3be4c

If this line is commented out, the test still passing.

Comment (luke-jr) [reply test/functional/wallet_multiwallet.py:120]: ???

Comment (hebasto) [reply test/functional/wallet_multiwallet.py:120]: If initial state is changed, i.e., the `no_access` directory actually has access, the test should fail, right?

Comment (luke-jr) [reply test/functional/wallet_multiwallet.py:120]: No, it's just there to make sure it doesn't break things. We're testing bitcoind, not the OS...

(And in fact, chmod won't work to deny read access on Windows)

**Review by hebasto — APPROVED** [review]
ACK 44c0bc31f4fadc932877d9764402644c84f3be4c

**Review by hebasto — APPROVED** [review]
re-ACK 24d2d3341d07509ad3f37bb6f130446ad20ac807, rebased only since my [previous](https://github.com/bitcoin/bitcoin/pull/19502#pullrequestreview-520552944) review.

File: src/wallet/walletutil.cpp [hunk 3]
```diff
@@ -49,26 +49,31 @@ std::vector<fs::path> ListWalletDir()
...
+                    // Found top-level btree file not called wallet.dat. Current bitcoin
+                    // software will never create these files but will allow them to be
+                    // opened in a shared database environment for backwards compatibility.
+                    // Add it to the list of available wallets.
+                    paths.emplace_back(path);
+                }
             }
+        } catch (const std::exception& e) {
+            LogPrintf("%s: Error scanning %s: %s\n", __func__, it->path().string(), e.what());
+            it.no_push();
```
Comment (promag) [inline src/wallet/walletutil.cpp:76]: nice, TIL

**Review by promag — COMMENTED** [review]
Tested ACK 24d2d3341d07509ad3f37bb6f130446ad20ac807, test change fails on master.

nit, could squash.

**Review by meshcollider — COMMENTED** [review]
utACK 24d2d3341d07509ad3f37bb6f130446ad20ac807

**Pull request merged**